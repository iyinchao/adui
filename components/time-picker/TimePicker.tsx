import React, {
  forwardRef,
  useContext,
  useImperativeHandle,
  useRef,
  useState,
  useMemo,
} from "react"
import PropTypes from "prop-types"
import classNames from "classnames"
import { ConfigContext, getComputedSize } from "../config-provider"
import Icon from "../icon"
import Input from "../input"
import Popover, { IPopoverProps } from "../popover"
import TimeSelect from "./TimeSelect"
import "./style"
import {
  transformTimeWithSeconds,
  extractHMSFromTime,
  isLegalTimeString,
  isTimeAfter,
  isTimeBefore,
} from "./core"

const prefix = "adui-time"
const noop = () => {}

export interface ITimePickerProps {
  [key: string]: any
  /**
   * 附加类名
   */
  className?: string
  /**
   * 是否在选择后，消失 Popover
   */
  closeOnSelect?: boolean
  /**
   * 内部驱动，默认选择时间
   */
  defaultValue?: null | string
  /**
   * 是否禁用
   */
  disabled?: boolean
  /**
   * 禁止选择的小时
   */
  disabledHours?: (hour?: string | null) => boolean | void
  /**
   * 禁止选择的分钟
   */
  disabledMinutes?: (minute?: string | null) => boolean | void
  /**
   * 禁止选择的秒
   */
  disabledSeconds?: (second?: string | null) => boolean | void
  /**
   * 设置输入框类型
   */
  intent?: "normal" | "primary" | "success" | "warning" | "danger"
  /**
   * 可选时间的最大值
   */
  maxTime?: string | null
  /**
   * 可选时间的最小值
   */
  minTime?: string | null
  /**
   * 时间改变的 handler
   */
  onChange?: (value: string) => void
  /**
   * 判断是否是整数
   */
  onlyHour?: boolean
  /**
   * 占位字符
   */
  placeholder?: string
  /**
   * 传入 Popover 的 props
   */
  popoverProps?: IPopoverProps
  /**
   * 20221216: 支持秒的选择
   */
  secondsAvailable?: boolean
  /**
   * 设置尺寸
   */
  size?: "mini" | "small" | "medium" | "large"
  /**
   * 外部控制：当前时间
   */
  value?: null | string
}

/**
 * 时间选择器用于了解当前已选时间，选择或输入需要的时间、时间范围等。
 */
const TimePicker: React.ForwardRefExoticComponent<
  ITimePickerProps & React.RefAttributes<any>
> = forwardRef(
  (
    {
      className,
      closeOnSelect,
      defaultValue,
      disabled,
      disabledHours,
      disabledMinutes,
      disabledSeconds,
      maxTime,
      minTime,
      onChange,
      onlyHour,
      placeholder,
      popoverProps,
      secondsAvailable = false,
      size: sizeProp,
      value,
      ...otherProps
    }: ITimePickerProps,
    ref
  ) => {
    const getInitialState = () => {
      const valueState = value !== null ? value : defaultValue
      const { hour, minute, second } = extractHMSFromTime(valueState)
      return {
        hour,
        inputValue: valueState || "",
        minute: onlyHour ? "00" : minute,
        second: secondsAvailable ? second : "00",
        prevValueProp: valueState || "",
        visible: false,
      }
    }
    const initialState = useMemo(getInitialState, [])
    const [inputValue, setInputValue] = useState(initialState.inputValue)
    const [hour, setHour] = useState(initialState.hour)
    const [minute, setMinute] = useState(initialState.minute)
    const [second, setSecond] = useState(initialState.second)
    const [prevValueProp, setPrevValueProp] = useState(
      initialState.prevValueProp
    )
    const [visible, setVisible] = useState(initialState.visible)

    if (value !== null && prevValueProp !== value) {
      const {
        hour: newHour,
        minute: newMinute,
        second: newSecond,
      } = extractHMSFromTime(value)

      setInputValue(value || "")
      setHour(newHour)
      setMinute(onlyHour ? "00" : newMinute)
      setSecond(secondsAvailable ? newSecond : "00")
      setPrevValueProp(value || "")
    }

    const inputRef = useRef<any>(null)

    const { size: sizeContext } = useContext(ConfigContext)

    const size = getComputedSize(sizeProp, sizeContext)

    const handleClick = (valueNew: string, type: "hour" | "minute") => {
      const valueProp = value

      let inputValueNow = ""
      if (secondsAvailable) {
        if (type === "hour") {
          inputValueNow = `${valueNew}:${minute || ""}:${second || ""}`
        } else if (type === "minute") {
          inputValueNow = `${hour || ""}:${valueNew}:${second || ""}`
        } else {
          inputValueNow = `${hour || ""}:${minute || ""}:${valueNew}`
        }
      } else {
        inputValueNow =
          type === "hour"
            ? `${valueNew}:${minute || ""}`
            : `${hour || ""}:${valueNew}`
      }

      if (maxTime && isTimeAfter(inputValueNow, maxTime)) {
        inputValueNow = secondsAvailable
          ? transformTimeWithSeconds(maxTime)
          : maxTime
      } else if (minTime && isTimeBefore(inputValueNow, minTime)) {
        inputValueNow = secondsAvailable
          ? transformTimeWithSeconds(minTime)
          : minTime
      }

      if (valueProp === null) {
        const valueFinal = {
          inputValueNow,
          ...extractHMSFromTime(inputValueNow),
        }
        setHour(valueFinal.hour)
        setMinute(valueFinal.minute)
        setSecond(valueFinal.second)
        setInputValue(valueFinal.inputValueNow)
      }

      if (isLegalTimeString({ time: inputValue, onlyHour })) {
        if (onChange) {
          onChange(inputValueNow)
        }
        if (closeOnSelect) {
          setTimeout(() => {
            setVisible(false)
          }, 120)
        }
      }
    }

    const handleVisibleChange = (visibleNew: boolean) => {
      const { input: inputElement } = inputRef.current
      if (disabled) {
        return
      }
      setTimeout(() => {
        const { activeElement } = document
        if (visibleNew || (!visibleNew && inputElement !== activeElement)) {
          const withSeconds =
            hour && minute && second ? `${hour}:${minute}:${second}` : ""
          const noSeconds = hour && minute ? `${hour}:${minute}` : ""
          const newInputValue = secondsAvailable ? withSeconds : noSeconds
          if (!visibleNew && inputValue !== newInputValue) {
            setInputValue(newInputValue)
          }
          setVisible(visibleNew)
        }
      }, 0)
    }

    const handleInputChange = ({
      target: { value: inputValueNew },
    }: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(inputValueNew)
      if (
        inputValueNew === "" ||
        (isLegalTimeString({
          time: inputValueNew,
          onlyHour,
        }) &&
          !isTimeAfter(inputValueNew, maxTime) &&
          !isTimeBefore(inputValueNew, minTime))
      ) {
        const {
          hour: hourNew,
          minute: minuteNew,
          second: secondNew,
        } = extractHMSFromTime(inputValueNew)
        if (
          !(disabledHours && disabledHours(hourNew)) &&
          !(disabledMinutes && disabledMinutes(minuteNew)) &&
          !(disabledSeconds && disabledSeconds(secondNew))
        ) {
          if (onChange) {
            onChange(inputValueNew)
          }
          if (value === null) {
            setHour(hourNew)
            setMinute(onlyHour ? "00" : minuteNew)
            setSecond(secondsAvailable ? secondNew : "00")
          }
        }
      }
    }

    const handleInputFocus = () => {
      if (!visible) {
        setVisible(true)
      }
    }

    const handleInputKeyDown = ({
      key,
    }: React.KeyboardEvent<HTMLInputElement>) => {
      const keys = ["Tab", "Escape", "Enter"]
      if (keys.includes(key)) {
        setVisible(false)
        inputRef.current.input.blur()
      }
    }

    useImperativeHandle(ref, () => ({
      input: inputRef.current,
      handleVisibleChange,
    }))

    const classSet = classNames(
      className,
      `${prefix}-input`,
      `${prefix}-${size}`,
      {
        [`${prefix}-secondsAvailable`]: secondsAvailable,
      }
    )

    return (
      <Popover
        arrowed={false}
        onVisibleChange={handleVisibleChange}
        popup={
          <div>
            <TimeSelect
              onChange={handleClick}
              currentHour={hour ? parseInt(hour, 10) : null}
              disabledHours={disabledHours}
              maxTime={maxTime}
              minTime={minTime}
              onlyHour={onlyHour}
              selectedValue={hour}
              size={size}
              type="hour"
            />
            {!onlyHour && (
              <TimeSelect
                disabledMinutes={disabledMinutes}
                onChange={handleClick}
                currentHour={hour ? parseInt(hour, 10) : null}
                maxTime={maxTime}
                minTime={minTime}
                selectedValue={minute}
                size={size}
                type="minute"
              />
            )}
            {secondsAvailable && (
              <TimeSelect
                disabledSeconds={disabledSeconds}
                onChange={handleClick}
                currentHour={hour ? parseInt(hour, 10) : null}
                currentMinute={minute ? parseInt(minute, 10) : null}
                maxTime={maxTime}
                minTime={minTime}
                selectedValue={second}
                size={size}
                type="second"
              />
            )}
          </div>
        }
        trigger="click"
        visible={visible}
        {...popoverProps}
      >
        <Input
          className={classSet}
          cleaveOptions={{
            blocks: secondsAvailable ? [2, 2, 2] : [2, 2],
            delimiter: ":",
          }}
          disabled={disabled}
          rightElement={<Icon icon="time-outlined" />}
          size={size}
          value={inputValue}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          onKeyDown={handleInputKeyDown}
          ref={inputRef}
          {...otherProps}
        />
      </Popover>
    )
  }
)

TimePicker.displayName = "TimePicker"

TimePicker.propTypes = {
  /**
   * 附加类名
   */
  className: PropTypes.string,
  /**
   * 是否在选择后，消失 Popover
   */
  closeOnSelect: PropTypes.bool,
  /**
   * 内部驱动，默认选择时间
   */
  defaultValue: PropTypes.string,
  /**
   * 是否禁用
   */
  disabled: PropTypes.bool,
  /**
   * 禁止选择的小时
   */
  disabledHours: PropTypes.func,
  /**
   * 禁止选择的分钟
   */
  disabledMinutes: PropTypes.func,
  /**
   * 禁止选择的秒
   */
  disabledSeconds: PropTypes.func,
  /**
   * 设置输入框类型
   */
  intent: PropTypes.oneOf([
    "normal",
    "primary",
    "success",
    "warning",
    "danger",
  ]),
  /**
   * 可选时间的最大值
   */
  maxTime: PropTypes.string,
  /**
   * 可选时间的最小值
   */
  minTime: PropTypes.string,
  /**
   * 时间改变的 handler
   */
  onChange: PropTypes.func,
  /**
   * 判断是否是整数
   */
  onlyHour: PropTypes.bool,
  /**
   * 占位字符
   */
  placeholder: PropTypes.string,
  /**
   * 设置尺寸
   */
  size: PropTypes.oneOf(["mini", "small", "medium", "large"]),
  /**
   * 外部控制：当前时间
   */
  value: PropTypes.string,
}

TimePicker.defaultProps = {
  className: "",
  closeOnSelect: false,
  defaultValue: null,
  disabled: false,
  disabledHours: noop,
  disabledMinutes: noop,
  disabledSeconds: noop,
  intent: "normal",
  maxTime: "24:00",
  minTime: null,
  onChange: noop,
  onlyHour: false,
  placeholder: "请选择",
  secondsAvailable: false,
  size: "small",
  value: null,
}

export default TimePicker
