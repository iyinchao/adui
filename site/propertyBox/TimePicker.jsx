import React from "react"
import styles from "./components/propertyBox.scss"
import CodeBox from "./components/CodeBox"
import { Button, Switch, TimePicker } from "../../components"
import { SIZES } from "./consts"

export default class PropertyBox extends React.Component {
  state = {
    onlyHour: false,
    disabled: false,
    size: "small",
    theme: false,
  }

  render() {
    const { onlyHour, disabled, size, theme } = this.state

    const codes = `<TimePicker
  ${disabled ? `disabled` : ""}
  ${onlyHour ? `onlyHour` : ""}
  ${size === "small" ? "" : `size="${size}"`}
  ${theme ? `theme="light"` : ""}
/>`

    const d = new Date()
    const hour = d.getHours()
    const minute = d.getMinutes()

    return (
      <div className={styles.wrapper}>
        <div className={styles.property}>
          <div className={styles.left}>
            {onlyHour ? (
              <TimePicker
                key="0"
                defaultValue={`${hour < 10 ? `0${hour}` : hour}:00`}
                disabled={disabled}
                size={size}
                onlyHour
                theme={theme ? "light" : null}
              />
            ) : (
              <TimePicker
                key="1"
                defaultValue={`${hour < 10 ? `0${hour}` : hour}:${
                  minute < 10 ? `0${minute}` : minute
                }`}
                disabled={disabled}
                size={size}
                theme={theme ? "light" : null}
              />
            )}
          </div>
          <div className={styles.right}>
            <div className={styles.rightTitle}>可配置项</div>
            <dl className={styles.controls}>
              <dt className={styles.controlTitle}>尺寸</dt>
              <dd className={styles.controlContent}>
                <Button.Group banner>
                  {SIZES.map(({ value, text }) => (
                    <Button
                      key={value}
                      active={value === size}
                      onClick={() => this.setState({ size: value })}
                    >
                      {text}
                    </Button>
                  ))}
                </Button.Group>
              </dd>
            </dl>
            <dl className={styles.controls}>
              <div className={styles.switchControl}>
                <Switch
                  checkedText="整点选择"
                  unCheckedText="整点选择"
                  checked={onlyHour}
                  onChange={(param) => this.setState({ onlyHour: param })}
                />
              </div>
              <div className={styles.switchControl}>
                <Switch
                  checkedText="禁用态"
                  unCheckedText="禁用态"
                  checked={disabled}
                  onChange={(param) => this.setState({ disabled: param })}
                />
              </div>
              <div className={styles.switchControl}>
                <Switch
                  checkedText="轻量风格"
                  unCheckedText="轻量风格"
                  checked={theme}
                  onChange={(param) => this.setState({ theme: param })}
                />
              </div>
            </dl>
          </div>
        </div>
        <CodeBox codes={codes} />
      </div>
    )
  }
}
