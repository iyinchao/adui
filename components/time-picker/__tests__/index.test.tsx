// @ts-nocheck
import * as React from "react"
import TestRenderer from "react-test-renderer"
import { act } from "react-dom/test-utils"
import { mount } from "enzyme"
import async from "async"
import TimePicker from ".."

class TimePickerTest extends React.Component {
  state = {
    value: "09:00",
  }

  render() {
    const { value } = this.state
    return (
      <div>
        <TimePicker
          closeOnSelect
          value={value}
          onChange={(val) => this.setState({ value: val })}
          minTime="08:40"
          maxTime="09:20"
          ref={(node) => {
            if (node) {
              this.picker = node
            }
            return null
          }}
        />
      </div>
    )
  }
}

function timeout(ms: number) {
  return (done: () => void) => {
    setTimeout(done, ms)
  }
}

describe("TimePicker", () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it("是否正确渲染", () => {
    const wrapper = TestRenderer.create(<TimePicker />)
    expect(wrapper).toMatchSnapshot()
  })

  it("各种情况快照核对", () => {
    const wrapper0 = TestRenderer.create(
      <TimePicker size="mini" theme="light" onlyHour disabled />
    )
    expect(wrapper0).toMatchSnapshot()
  })

  it("输入事件", () => {
    const wrapper = mount(<TimePicker />)
    const input = wrapper.find("input")
    input.simulate("focus")
    expect(wrapper.find("Popover").props().visible).toEqual(true)
    input.simulate("change", { target: { value: " " } })
    input.simulate("change", { target: { value: "16:07" } })
    expect(wrapper.find("Input").props().value).toEqual("16:07")
    input.simulate("keydown", { keyCode: 13 })
    expect(wrapper.find("Popover").props().visible).toEqual(false)
  })

  it("点击事件", () => {
    const wrapper = mount(<TimePickerTest />)

    expect(wrapper.find("Popover").props().visible).toEqual(false)

    wrapper.find("input").simulate("focus")
    expect(wrapper.find("Popover").props().visible).toEqual(true)

    act(() => {
      wrapper.find(".adui-time-item").at(0).simulate("click")
      jest.advanceTimersByTime(50)
      wrapper.update()
      expect(wrapper.find("Input").props().value).toEqual("08:40")

      wrapper.find(".adui-time-item").at(42).simulate("click")
      wrapper.update()
      jest.advanceTimersByTime(50)
      expect(wrapper.find("Input").props().value).toEqual("08:40")

      wrapper.find("input").simulate("change", { target: { value: "07:40" } })
      wrapper.find("input").simulate("change", { target: { value: "08:4" } })
      wrapper.update()
      expect(wrapper.find("Input").props().value).toEqual("08:40")
    })

    // act(() => {
    //   wrapper
    //     .find(".adui-time-item")
    //     .at(1)
    //     .simulate("click")
    //   wrapper.update()
    //   jest.advanceTimersByTime(350)
    //   expect(wrapper.find("Input").props().value).toEqual("09:20")
    //   done()
    // })

    // act(() => {
    //   wrapper
    //     .find(".adui-time-item")
    //     .at(1)
    //     .simulate("click")
    //   wrapper.update()
    //   expect(wrapper.find("Input").props().value).toEqual("09:20")
    // })

    // act(() => {
    //   wrapper
    //     .find(".adui-time-item")
    //     .at(0)
    //     .simulate("click")
    //   wrapper.update()
    //   expect(wrapper.find("Input").props().value).toEqual("08:40")
    // })
  })

  it("内部驱动", () => {
    const wrapper = mount(<TimePicker defaultValue="09:20" />)
    const input = wrapper.find("input")
    input.simulate("focus")
    expect(wrapper.find("Popover").props().visible).toEqual(true)
    wrapper.find(".adui-time-item").at(0).simulate("click")
    expect(wrapper.find("Input").props().value).toEqual("00:20")
  })

  it("外部控制", () => {
    const wrapper = mount(<TimePicker value="09:00" onlyHour />)
    const input = wrapper.find("input")
    input.simulate("focus")
    expect(wrapper.find("Popover").props().visible).toEqual(true)
    wrapper.find(".adui-time-item").at(0).simulate("click")
    expect(wrapper.find("Input").props().value).toEqual("09:00")
  })

  it("测试禁用", () => {
    const wrapper = mount(<TimePicker disabled />)
    const input = wrapper.find("input")
    input.simulate("click")
    expect(wrapper.find("Popover").props().visible).toEqual(false)
  })

  it("测试最大时间和最小时间", () => {
    const wrapper = mount(
      <TimePicker minTime="09:00" maxTime="09:20" defaultValue="09:10" />
    )
    const input = wrapper.find("input")
    input.simulate("change", { target: { value: "09:30" } })
    input.simulate("focus")
    expect(wrapper.find("Popover").props().visible).toEqual(true)
  })
})
