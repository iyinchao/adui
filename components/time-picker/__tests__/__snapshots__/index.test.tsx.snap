// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker 各种情况快照核对 1`] = `
<div
  className="adui-time-input adui-time-mini adui-input-wrapper adui-input-mini adui-input-light adui-input-disabled adui-input-normal"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={Object {}}
>
  <input
    autoComplete="off"
    autoFocus={false}
    className="adui-input-base"
    disabled={true}
    id={null}
    name={null}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="请输入"
    required={false}
    style={
      Object {
        "paddingLeft": undefined,
        "paddingRight": undefined,
      }
    }
    type={null}
    value=""
  />
  <div
    className="adui-input-right adui-input-right_icon"
    onClick={[Function]}
    role="none"
  >
    <svg
      className="adui-icon-base"
      data-icon="time-outlined"
      data-interactive={false}
      fill="var(--gray-700)"
      height={18}
      onClick={[Function]}
      viewBox="0 0 18 18"
      width={18}
    >
      <path
        d="M9.75 9.18301L12.4641 10.75L11.7141 12.049L8.25 10.049L8.27831 10H8.25V5H9.75V9.18301ZM9 16C5.13401 16 2 12.866 2 9C2 5.13401 5.13401 2 9 2C12.866 2 16 5.13401 16 9C16 12.866 12.866 16 9 16ZM9 14.5C12.0376 14.5 14.5 12.0376 14.5 9C14.5 5.96243 12.0376 3.5 9 3.5C5.96243 3.5 3.5 5.96243 3.5 9C3.5 12.0376 5.96243 14.5 9 14.5Z"
        fillRule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`TimePicker 是否正确渲染 1`] = `
<div
  className="adui-time-input adui-time-small adui-input-wrapper adui-input-small adui-input-normal"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={Object {}}
>
  <input
    autoComplete="off"
    autoFocus={false}
    className="adui-input-base"
    disabled={false}
    id={null}
    name={null}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="请输入"
    required={false}
    style={
      Object {
        "paddingLeft": undefined,
        "paddingRight": undefined,
      }
    }
    type={null}
    value=""
  />
  <div
    className="adui-input-right adui-input-right_icon"
    onClick={[Function]}
    role="none"
  >
    <svg
      className="adui-icon-base"
      data-icon="time-outlined"
      data-interactive={false}
      fill="var(--gray-700)"
      height={18}
      onClick={[Function]}
      viewBox="0 0 18 18"
      width={18}
    >
      <path
        d="M9.75 9.18301L12.4641 10.75L11.7141 12.049L8.25 10.049L8.27831 10H8.25V5H9.75V9.18301ZM9 16C5.13401 16 2 12.866 2 9C2 5.13401 5.13401 2 9 2C12.866 2 16 5.13401 16 9C16 12.866 12.866 16 9 16ZM9 14.5C12.0376 14.5 14.5 12.0376 14.5 9C14.5 5.96243 12.0376 3.5 9 3.5C5.96243 3.5 3.5 5.96243 3.5 9C3.5 12.0376 5.96243 14.5 9 14.5Z"
        fillRule="evenodd"
      />
    </svg>
  </div>
</div>
`;
